The aim of this project is to clean the dataset and perform some initial analysis on it. While working on this project, you'll become familiar with some of the unique benefits the JupyterLab environment has to offer for pandas.

In [2]:
import pandas as pd
import numpy as np
import re

In [3]:
autos= pd.read_csv("autos.csv", encoding="Latin-1")

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

,dateCrawled,name,seller,offerType,price,abtest,vehicleType,yearOfRegistration,gearbox,powerPS,model,kilometer,monthOfRegistration,fuelType,brand,notRepairedDamage,dateCreated,nrOfPictures,postalCode,lastSeen
0,2016-03-24 11:52:17,Golf_3_1.6,privat,Angebot,480,test,NaN,1993,manuell,0,golf,150000,0,benzin,volkswagen,NaN,2016-03-24 00:00:00,0,70435,2016-04-07 03:16:57
1,2016-03-24 10:58:45,A5_Sportback_2.7_Tdi,privat,Angebot,18300,test,coupe,2011,manuell,190,NaN,125000,5,diesel,audi,ja,2016-03-24 00:00:00,0,66954,2016-04-07 01:46:50
2,2016-03-14 12:52:21,"Jeep_Grand_Cherokee_""Overland""",privat,Angebot,9800,test,suv,2004,automatik,163,grand,125000,8,diesel,jeep,NaN,2016-03-14 00:00:00,0,90480,2016-04-05 12:47:46
3,2016-03-17 16:54:04,GOLF_4_1_4__3TÜRER,privat,Angebot,1500,test,kleinwagen,2001,manuell,75,golf,150000,6,benzin,volkswagen,nein,2016-03-17 00:00:00,0,91074,2016-03-17 17:40:17
4,2016-03-31 17:25:20,Skoda_Fabia_1.4_TDI_PD_Classic,privat,Angebot,3600,test,kleinwagen,2008,manuell,69,fabia,90000,7,diesel,skoda,nein,2016-03-31 00:00:00,0,60437,2016-04-06 10:17:21


The data set contains 20 columns with 371,258 rows of information most of which are string values.Some columns have null values but none have more than 20% null values

DATA CLEANING

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#Rename specific columns
autos= autos.rename(columns={"yearOfRegistration":"registration_year", "monthOfRegistration" :"registration_month", "notRepairedDamage" : "unrepaired_damage","dateCreated" : "ad_created"})

In [7]:
#Convert remaining columns from camelcase to snakecase
autos.columns= [re.sub(r'([a-z])([A-Z])', r'\1_\2', col).lower() for col in autos.columns]
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [8]:
autos.head()

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,kilometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,nr_of_pictures,postal_code,last_seen
0,2016-03-24 11:52:17,Golf_3_1.6,privat,Angebot,480,test,NaN,1993,manuell,0,golf,150000,0,benzin,volkswagen,NaN,2016-03-24 00:00:00,0,70435,2016-04-07 03:16:57
1,2016-03-24 10:58:45,A5_Sportback_2.7_Tdi,privat,Angebot,18300,test,coupe,2011,manuell,190,NaN,125000,5,diesel,audi,ja,2016-03-24 00:00:00,0,66954,2016-04-07 01:46:50
2,2016-03-14 12:52:21,"Jeep_Grand_Cherokee_""Overland""",privat,Angebot,9800,test,suv,2004,automatik,163,grand,125000,8,diesel,jeep,NaN,2016-03-14 00:00:00,0,90480,2016-04-05 12:47:46
3,2016-03-17 16:54:04,GOLF_4_1_4__3TÜRER,privat,Angebot,1500,test,kleinwagen,2001,manuell,75,golf,150000,6,benzin,volkswagen,nein,2016-03-17 00:00:00,0,91074,2016-03-17 17:40:17
4,2016-03-31 17:25:20,Skoda_Fabia_1.4_TDI_PD_Classic,privat,Angebot,3600,test,kleinwagen,2008,manuell,69,fabia,90000,7,diesel,skoda,nein,2016-03-31 00:00:00,0,60437,2016-04-06 10:17:21


I have changed some of the column names so that they can be more descriptive. I have also converted the naming convention from camelcase to snailcase to improve readability.

In [9]:
autos.describe(include='all')

,date_crawled,name,seller,offer_type,price,abtest,vehicle_type,registration_year,gearbox,power_ps,model,kilometer,registration_month,fuel_type,brand,unrepaired_damage,ad_created,nr_of_pictures,postal_code,last_seen
count,371528,371528,371528,371528,3.715280e+05,371528,333659,371528.000000,351319,371528.000000,351044,371528.000000,371528.000000,338142,371528,299468,371528,371528.0,371528.00000,371528
unique,280500,233531,2,2,NaN,2,8,NaN,2,NaN,251,NaN,NaN,7,40,2,114,NaN,NaN,182806
top,2016-03-24 14:49:47,Ford_Fiesta,privat,Angebot,NaN,test,limousine,NaN,manuell,NaN,golf,NaN,NaN,benzin,volkswagen,nein,2016-04-03 00:00:00,NaN,NaN,2016-04-06 13:45:54
freq,7,657,371525,371516,NaN,192585,95894,NaN,274214,NaN,30070,NaN,NaN,223857,79640,263182,14450,NaN,NaN,17
mean,NaN,NaN,NaN,NaN,1.729514e+04,NaN,NaN,2004.577997,NaN,115.549477,NaN,125618.688228,5.734445,NaN,NaN,NaN,NaN,0.0,50820.66764,NaN
std,NaN,NaN,NaN,NaN,3.587954e+06,NaN,NaN,92.866598,NaN,192.139578,NaN,40112.337051,3.712412,NaN,NaN,NaN,NaN,0.0,25799.08247,NaN
min,NaN,NaN,NaN,NaN,0.000000e+00,NaN,NaN,1000.000000,NaN,0.000000,NaN,5000.000000,0.000000,NaN,NaN,NaN,NaN,0.0,1067.00000,NaN
25%,NaN,NaN,NaN,NaN,1.150000e+03,NaN,NaN,1999.000000,NaN,70.000000,NaN,125000.000000,3.000000,NaN,NaN,NaN,NaN,0.0,30459.00000,NaN
50%,NaN,NaN,NaN,NaN,2.950000e+03,NaN,NaN,2003.000000,NaN,105.000000,NaN,150000.000000,6.000000,NaN,NaN,NaN,NaN,0.0,49610.00000,NaN
75%,NaN,NaN,NaN,NaN,7.200000e+03,NaN,NaN,2008.000000,NaN,150.000000,NaN,150000.000000,9.000000,NaN,NaN,NaN,NaN,0.0,71546.00000,NaN


The seller and offer type columns almost have the same reoccuring value and therefore need to be dropped. Further investigation needs to be done to the nr_of_pictures column as it has no values

In [10]:
autos["nr_of_pictures"].value_counts()

nr_of_pictures
0    371528
Name: count, dtype: int64

We will drop the three columns

In [11]:
autos= autos.drop(["nr_of_pictures","seller","offer_type"], axis= 1)

In [12]:
autos=autos.rename(columns={"kilometer" : "odometer_km"})

Exploring Odometer and Price

In [13]:
autos['odometer_km'].value_counts()


odometer_km
150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: count, dtype: int64

Most cars have a higher mileage and the rounded figures tell us that this was preset value.

In [14]:
print(autos['price'].describe())
autos['price'].value_counts()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


price
0        10778
500       5670
1500      5394
1000      4649
1200      4594
         ...  
11333        1
1898         1
378          1
21856        1
408          1
Name: count, Length: 5597, dtype: int64

In order to maintain accuracy, we will drop the 10,778 rows where cars have the price as 0. the highest price of a car is roughly 2.5 billion which seems too high and therefore requires further investiagtion.

In [15]:
autos["price"].value_counts().sort_index(ascending= False).head(20)

price
2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: count, dtype: int64

In [16]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

price
0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: count, dtype: int64

We have a couple of listings with the price set at zero that need to be dropped. however, we will keep values from 1 being that eBay is an auction site. For the higher prices we will keep values from 3.5 million and below because prices increase steadily up until that number where they spike to unrealistic values.

In [17]:
autos=autos[autos["price"].between(1,3900000)]
autos["price"].describe()

count    3.606980e+05
mean     6.068401e+03
std      1.818741e+04
min      1.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.895000e+06
Name: price, dtype: float64

Exploring the Date Columns

In [18]:
autos.dtypes

date_crawled          object
name                  object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

In [19]:
autos[['date_crawled','ad_created','last_seen']][0:5]

,date_crawled,ad_created,last_seen
0,2016-03-24 11:52:17,2016-03-24 00:00:00,2016-04-07 03:16:57
1,2016-03-24 10:58:45,2016-03-24 00:00:00,2016-04-07 01:46:50
2,2016-03-14 12:52:21,2016-03-14 00:00:00,2016-04-05 12:47:46
3,2016-03-17 16:54:04,2016-03-17 00:00:00,2016-03-17 17:40:17
4,2016-03-31 17:25:20,2016-03-31 00:00:00,2016-04-06 10:17:21


The columns date_created, last_seen and ad_created have data on dates but are stored as strings. We therefore need to convert them to datetime format.

In [20]:
(autos['date_crawled']
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index())

date_crawled
2016-03-05    0.025545
2016-03-06    0.014480
2016-03-07    0.035653
2016-03-08    0.033466
2016-03-09    0.034112
2016-03-10    0.032648
2016-03-11    0.032775
2016-03-12    0.036244
2016-03-13    0.015781
2016-03-14    0.036327
2016-03-15    0.033424
2016-03-16    0.030200
2016-03-17    0.031655
2016-03-18    0.013119
2016-03-19    0.035268
2016-03-20    0.036399
2016-03-21    0.035695
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032798
2016-03-26    0.031969
2016-03-27    0.030230
2016-03-28    0.035060
2016-03-29    0.034123
2016-03-30    0.033529
2016-03-31    0.031874
2016-04-01    0.034150
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037630
2016-04-05    0.012786
2016-04-06    0.003127
2016-04-07    0.001619
Name: proportion, dtype: float64

The site was crawled daily for over a month from the month of March to April

In [21]:
(autos['ad_created']
    .str[:10]
    .value_counts(normalize= True, dropna=False)
    .sort_index())

ad_created
2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038999
2016-04-04    0.037738
2016-04-05    0.011619
2016-04-06    0.003119
2016-04-07    0.001555
Name: proportion, Length: 114, dtype: float64

Ads were created from the year 2014 to 2016 and it is not distributed uniformly across all months.

In [22]:
(autos['last_seen']
    .str[:10]
    .value_counts(normalize=True, dropna= False)
    .sort_index())

last_seen
2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005201
2016-03-08    0.007940
2016-03-09    0.009823
2016-03-10    0.011458
2016-03-11    0.012953
2016-03-12    0.023241
2016-03-13    0.008409
2016-03-14    0.012174
2016-03-15    0.016327
2016-03-16    0.016415
2016-03-17    0.028703
2016-03-18    0.006892
2016-03-19    0.016332
2016-03-20    0.019889
2016-03-21    0.020031
2016-03-22    0.020510
2016-03-23    0.018018
2016-03-24    0.019163
2016-03-25    0.018997
2016-03-26    0.015955
2016-03-27    0.016720
2016-03-28    0.022190
2016-03-29    0.023285
2016-03-30    0.023721
2016-03-31    0.024239
2016-04-01    0.023898
2016-04-02    0.024971
2016-04-03    0.025320
2016-04-04    0.025531
2016-04-05    0.126968
2016-04-06    0.218942
2016-04-07    0.130422
Name: proportion, dtype: float64

The last seen dates are distrubuted uniformly in the year 2016 throughout the month of March and beginning of April

In [23]:
autos["registration_year"].describe()

count    360698.000000
mean       2004.451877
std          82.110038
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

There are 360,698 entries of car registration. However the maximum value of 9999 and minimum value of 1000 call for further investigation as its not possible to have these as registration years.

DEALING WITH INCORRECT REGISTRATION YEAR DATA

A car can't be registered after the first listing was made, therefore any car registered after the year 2016 should be filtered out. For this analysis, we will use our earliest year as 1900

In [24]:
(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]

np.float64(0.03875264071328369)

Vehicles that fall outside of the range of 1900 and 2016 are less than 4% thereofre we can drop those columns

In [25]:
autos= autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000    0.066702
1999    0.063544
2005    0.062670
2006    0.057704
2001    0.056948
2003    0.056547
2004    0.056169
2002    0.054280
2007    0.050490
1998    0.049691
Name: proportion, dtype: float64

Most vehicles are registered between the year 1998 and 2007

EXPLORING PRICE BY BRAND

In [26]:
autos["brand"].value_counts(normalize= True)

brand
volkswagen        0.211675
bmw               0.109870
opel              0.106397
mercedes_benz     0.096833
audi              0.089536
ford              0.068912
renault           0.047508
peugeot           0.030148
fiat              0.025686
seat              0.018658
skoda             0.015684
mazda             0.015381
smart             0.014329
citroen           0.013948
nissan            0.013596
toyota            0.012930
hyundai           0.009971
sonstige_autos    0.009547
mini              0.009382
volvo             0.009146
mitsubishi        0.008234
honda             0.007531
kia               0.006913
suzuki            0.006365
alfa_romeo        0.006308
porsche           0.006259
chevrolet         0.005024
chrysler          0.003862
dacia             0.002495
jeep              0.002195
land_rover        0.002166
daihatsu          0.002160
subaru            0.002117
jaguar            0.001736
saab              0.001465
daewoo            0.001457
trabant           0.00

Volkswagen, BMW,Opel, Mercedes,Audi and Ford each individually account for 5% of the brands. Therefore we will limit our analysis to these listings.

In [27]:
brand_counts=autos["brand"].value_counts(normalize= True)
top_brands= brand_counts[brand_counts>.05].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In [28]:
brand_mean_prices={}

for brand in top_brands:
    top_brand_only=autos[autos["brand"]==brand]
    mean_price= top_brand_only["price"].mean()
    brand_mean_prices[brand]= int(mean_price)

brand_mean_prices

{'volkswagen': 5452,
 'bmw': 8573,
 'opel': 3028,
 'mercedes_benz': 8609,
 'audi': 9166,
 'ford': 3779}

While volkswagen accoints for a large percentage of sales, Audi has the highest price with a mean price of 9166. Opel offers the cheapest mean price of 3028 while Volkswagen has a mid price range of 5452. This explain why it has the highest sales as it is not too cheap or too expensive.

STORING AGGREGATE DATA IN A DATA FRAME

In [29]:
bmp_series= pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

,mean_price
volkswagen,5452
bmw,8573
opel,3028
mercedes_benz,8609
audi,9166
ford,3779


In [35]:
brand_mean_mileage= {}

for brand in top_brands:
    top_brands_only=autos[autos["brand"]==brand]
    mean_mileage= top_brands_only["odometer_km"].mean()
    brand_mean_mileage[brand]= int(mean_mileage)

mean_prices= pd.Series(brand_mean_prices).sort_values(ascending= False)
mean_mileage=pd.Series(brand_mean_mileage).sort_values(ascending=False)

brand_mean_mileage

{'volkswagen': 128387,
 'bmw': 132791,
 'opel': 128721,
 'mercedes_benz': 130566,
 'audi': 129440,
 'ford': 123663}

In [36]:
brand_info=pd.DataFrame(mean_mileage,columns=["mean_mileage"])
brand_info

,mean_mileage
bmw,132791
mercedes_benz,130566
audi,129440
opel,128721
volkswagen,128387
ford,123663


In [37]:
brand_info["mean_price"]= mean_prices
brand_info

,mean_mileage,mean_price
bmw,132791,8573
mercedes_benz,130566,8609
audi,129440,9166
opel,128721,3028
volkswagen,128387,5452
ford,123663,3779


The car mileage does not vary as much in comparison to the car prices. However, more expensive cars tend to have a higher mileage